hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -1154,31 +1184,26 @@ def example_deletion_with_block_lowering(self):
         example and every block not inside that example it tries deleting the
         example and modifying the block's value by one in either direction.
         """
-        i = 0
-        while i < len(self.shrink_target.blocks):
-            if not self.is_shrinking_block(i):
-                i += 1
-                continue
-
-            u, v = self.blocks[i].bounds
 
-            j = 0
-            while j < len(self.shrink_target.examples):
-                n = int_from_bytes(self.shrink_target.buffer[u:v])
-                if n == 0:
-                    break
-                ex = self.shrink_target.examples[j]
-                if ex.start < v or ex.length == 0:
-                    j += 1
-                    continue
-
-                buf = bytearray(self.shrink_target.buffer)
-                buf[u:v] = int_to_bytes(n - 1, v - u)
-                del buf[ex.start : ex.end]
-                if not self.incorporate_new_buffer(buf):
-                    j += 1
-
-            i += 1
+        for ex, block in self.each_pair(
+            base_collection=lambda: sorted(
+                self.shrink_target.examples, key=lambda ex: ex.depth
+            ),
+            dependent_collection=lambda ex: [
+                b
+                for b in self.blocks
```

Comment:
One thing I noticed is that this will tend to pair most examples with block 0, then pair most examples with block 1, and so on. That works, but isn't ideal for diversity now that we don't have the `%` to lean on any more.

(`each_pair_of_blocks` doesn't have this problem, because its iteration pattern is naturally “triangular”, so it gets `j` diversity for free.)

However, it occurred to me that if this is `for b in reversed(self.blocks)`, then we gain a lot more diversity, and we also prioritize example/block pairs that are closer to each other, which seems like a good ordering heuristic.

I haven't thought through the full implications of this, so there's a chance it's a bad idea, but it seems promising.